In [2]:
!pip install cvc5

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.5/13.5 MB 85.4 MB/s eta 0:00:00


# Aritmética Inteira e Real
A aritmética pode ser classificada em diferentes categorias dependendo do tipo de números e operações envolvidas.  
Neste notebook, exploraremos quatro tipos principais de aritmética frequentemente usadas em lógica matemática e computação.


## 1. Aritmética Linear Inteira (LIA)
A Aritmética Linear Inteira (LIA) lida com equações lineares cujas variáveis assumem apenas valores inteiros.  
Ela é muito utilizada em problemas de programação inteira, onde as soluções precisam ser números inteiros, como alocação de recursos e planejamento de produção.



### Exemplo 1:
Resolver a equação linear para valores inteiros:

Equação Única
$$
3x = 15
$$

onde $ x $ são inteiros.


In [21]:
from cvc5.pythonic import *
x = Int('x')
solve(3*x == 15)  # [x = 5]

[x = 5]


### Exemplo 2:
Resolver a equação linear para valores inteiros:

Duas variáveis
$$
x + y = 10  
$$
$$
2x - y = 5
$$

onde $x$, $y$ são inteiros.

In [22]:
x, y = Ints('x y')
solve(x + y == 10, 2*x - y == 5)  # [x = 5, y = 5]

[x = 5, y = 5]


### Exemplo 3:
Resolver a equação linear para valores inteiros:

Planejamento de Produção
Encontre valores inteiros onde:

$$
5a + 3b ≤ 30  
$$
$$
a ≥ 2
$$
$$
b ≥ 4  
$$

e maximizar
$$
4a + 5b
$$

onde $a$, $b$ são inteiros.

In [23]:
from cvc5.pythonic import *

a, b = Ints('a b')
max_profit = 0
best_solution = None

while True:
    solver = Solver()
    solver.add(5*a + 3*b <= 30, a >= 2, b >= 4)
    solver.add(4*a + 5*b > max_profit)  # Seek better solutions

    if solver.check() == sat:
        model = solver.model()
        current_profit = model.evaluate(4*a + 5*b).as_long()
        max_profit = current_profit
        best_solution = model
    else:
        break

print(f"Optimal solution: {best_solution}")
print(f"Maximum profit: {max_profit}")

Optimal solution: [a = 2, b = 6]
Maximum profit: 38


## 2. Aritmética Linear Real (LRA)
A Aritmética Linear Real (LRA) lida com equações lineares onde as variáveis podem assumir valores reais.  
Ela é amplamente utilizada em otimização, economia e engenharia.




### Exemplo 1:
Resolver a equação linear para valores inteiros:

Razão Simples
$$
2.5x = 10.0
$$

onde $ x $ são reais.


In [24]:
x = Real('x')
solve(2.5*x == 10.0)

[x = 4]


### Exemplo 2:
Resolver a equação linear para valores inteiros:

Problema de mistura
$$
0.3x + 0.6y = 12
$$
$$
x + y = 30
$$

onde $x$, $y$ são reais.

In [25]:
x, y = Reals('x y')
solve(0.3*x + 0.6*y == 12, x + y == 30)

[x = 20, y = 10]


### Exemplo 3:
Resolver a equação linear para valores inteiros:

Determinar se existe um plano de produção que satisfaça:
$$
1,8A + 2,4B + 1,2C ≤ 1200 \, (\text{Energia [MJ]})
$$

$$
0,9A + 1,5B + 2,1C ≤ 800 \, (\text{Matéria-prima [toneladas]})
$$

$$
2,5A + 1,8B + 3,0C ≤ 1500 \, (\text{Água [m³]})
$$

$$
A ≥ 50 \, (\text{Produto Mínimo A})
$$

$$
B ≥ 0,6C \, (\text{Proporção B/C do Produto})
$$

$$
C ≤ 2A \, (\text{C Máx. em relação a A})
$$

$$
A + B + C ≥ 300 \, (\text{Produção total})
$$


onde $A$, $B$ são reais.

In [26]:
from cvc5.pythonic import *

A, B, C = Reals('A B C')

solver = Solver()
solver.add(
    1.8*A + 2.4*B + 1.2*C <= 1200,
    0.9*A + 1.5*B + 2.1*C <= 800,
    2.5*A + 1.8*B + 3.0*C <= 1500,
    A >= 50,
    B >= 0.6*C,
    C <= 2*A,
    A + B + C >= 300
)

# check:
print("Basic feasibility:", solver.check())
if solver.check() == sat:
    model = solver.model()
    print(f"Sample solution: A={model[A]}, B={model[B]}, C={model[C]}")

solver.push()  # Temporary constraint
solver.add(B >= 450)
print("\nWith B≥450:", solver.check())
print("Basic feasibility:", solver.check())
if solver.check() == sat:
    model = solver.model()
    print(f"Sample solution: A={model[A]}, B={model[B]}, C={model[C]}")
solver.pop()  # Remove temporary constraint

Basic feasibility: sat
Sample solution: A=50, B=250, C=0

With B≥450: sat
Basic feasibility: sat
Sample solution: A=50, B=450, C=-200


## 3. Aritmética Não Linear Inteira (NIA)
A Aritmética Não Linear Inteira (NIA) trata de equações não lineares onde as variáveis assumem apenas valores inteiros.  
Este tipo de problema é mais desafiador devido à complexidade combinatória.





### Exemplo 1:
Resolver a Equação Diofantina para valores inteiros:

$$
x² + y² = z² + 20
$$
$$
x > y > 0
$$
$$
 z > 5
$$

onde $ x,y,z $ são inteiros.


In [27]:
from cvc5.pythonic import *

x, y, z = Ints('x y z')
solve(x > y, y > 0, z > 5, x*x + y*y == z*z + 20,)

[x = 10, y = 1, z = 9]


In [28]:
from cvc5.pythonic import *
x, y, z = Ints('x y z')
s = SolverFor('QF_NIA')  # Usando Linear Real Arithmetic

# Adicionando as restrições
s.add(x > y, y > 0, z > 5, x*x + y*y == z*z + 20)

print(s.model() if s.check() == sat else "unsat")

[x = 10, y = 1, z = 9]


### Exemplo 2:
Encontre fatores de 221 como
$$
a * b = 221
$$


onde $a$, $b$ são inteiros e primos.

In [29]:
a, b = Ints('a b')
solve(a > 1, b > 1, a * b == 221, a != b)

[a = 13, b = 17]


### Exemplo 3:
Resolver o Sistema não linear:

$$
x³ + y² = 250
$$

$$
x × y = 20
$$

$$
x > y ≥ 2
$$




onde $x$, $y$ são inteiros.

In [30]:
x, y = Ints('x y')
solve(x > y, y >= 2, x**3 + y**2 == 141, x * y == 20)

[x = 5, y = 4]


## 4. Aritmética Não Linear Real (NRA)
A Aritmética Não Linear Real (NRA) lida com equações não lineares onde as variáveis podem assumir valores reais.  
Ela é aplicada em áreas como física, economia e aprendizado de máquina.




### Exemplo 1:
Sistema Quadrático
Resolva para números reais:
$$
x² + y = 10
$$
$$
x * y = 12
$$

onde $ x, y $ são reais.


In [31]:
from cvc5.pythonic import *
x, y = Reals('x y')
s = SolverFor('QF_NRA')
s.add(x**2 + y == 10, x * y == 12)
s.add(x * y == 12)
print(s.model() if s.check() == sat else "unsat")

[x = (_ real_algebraic_number <1*x^3 + (-10*x) + 12, (3/2, 7/4)>), y = (_ real_algebraic_number <1*x^3 + (-10*x^2) + 144, (29/4, 15/2)>)]


In [32]:
from cvc5.pythonic import *
x, y = Reals('x y')
s = SolverFor('QF_NRA')
s.add(x**2 + y == 10, x * y == 12, x==2)
s.add(x * y == 12)
print(s.model() if s.check() == sat else "unsat")

[x = 2, y = 6]



Output: [x = 2, y = 6] or [x = -3, y = -4]

# Bitvectors
Bitvectors são sequências de bits (0s e 1s) usadas para representar dados de forma compacta e eficiente. Cada bit em um bitvector pode representar informações binárias, como estados de dispositivos ou flags. Operações lógicas e aritméticas, como AND, OR, e SHIFT, são realizadas em nível de bit, permitindo manipulação rápida e eficiente. Bitvectors são usados em áreas como criptografia, otimização de algoritmos e computação simbólica.

##Exemplo 1:
Representação de Dados Binários

In [33]:
from cvc5.pythonic import *

x = BitVec('x', 8)  # x é um bitvector de 8 bits
s = SolverFor('QF_BV')  # Solver para bitvectors
s.add(x * 5 == 1)  # Adicionando restrição para x * 5 == 1
result = s.check()  # Checando se a fórmula é satisfatível
print("result: ", result)
if result == sat:
    m = s.model()
    print("\n".join([str(k) + " : " + str(m[k]) for k in m]))

result:  sat
x : 205


o bitvector x é utilizado para representar um valor binário de 8 bits. A restrição x * 5 == 1 é uma operação simples em bitvectors, onde o solver tenta encontrar um valor para x que satisfaça essa equação. Esse exemplo demonstra como um bitvector pode ser manipulado de forma eficiente para representar e resolver operações binárias.

Esse tipo de operação é útil quando precisamos representar números inteiros de forma compacta e manipulá-los diretamente em nível de bits, especialmente em computação de baixo nível, criptografia e design de hardware.

##Exemplo 2:
 Implementação de Função de Valor Absoluto

In [34]:
from cvc5.pythonic import *
x = Const("x", BitVecSort(32))  # x é um bitvector de 32 bits
xrs = x >> 31  # Deslocamento de x 31 bits à direita
abs_ref = If(x < 0, -x, x)   # Implementação de abs()
abs1 = (x ^ xrs) - xrs  # Alternativa 1 para abs(x)
abs2 = (x + xrs) ^ xrs  # Alternativa 2 para abs(x)
abs3 = x - ((x << 1) & xrs)  # Alternativa 3 para abs(x)

prove(abs_ref == abs1)  # Verificando se abs_ref == abs1
prove(abs_ref == abs2)  # Verificando se abs_ref == abs2
prove(abs_ref == abs3)  # Verificando se abs_ref == abs3

proved
proved
proved


Este exemplo ilustra três maneiras diferentes de calcular o valor absoluto de um número representado como um bitvector de 32 bits.

A operação de deslocamento de bits (>> e <<) é usada para manipular diretamente os bits do valor de x. As três alternativas (abs1, abs2, abs3) tentam calcular o valor absoluto de x usando diferentes truques bit a bit, sem a necessidade de uma operação condicional como if.

O prove é utilizado para verificar a equivalência entre a implementação convencional (abs_ref) e as alternativas propostas.

Esse exemplo mostra como operações aritméticas e de manipulação de bits podem ser usadas para calcular funções matemáticas de forma eficiente. Em sistemas embarcados e computação de baixo nível, essas operações são fundamentais para realizar cálculos rápidos e eficientes sem sobrecarregar o sistema com funções complexas.

##Exemplo 3:
Prova de Propriedade de um Algoritmo

In [35]:
prove(abs_ref == abs1)  # Verificando se abs_ref == abs1
prove(abs_ref == abs2)  # Verificando se abs_ref == abs2
prove(abs_ref == abs3)  # Verificando se abs_ref == abs3

proved
proved
proved


Neste exemplo, as provas são realizadas para garantir que as diferentes implementações de valor absoluto (abs1, abs2, abs3) sejam equivalentes à implementação tradicional (abs_ref).
Isso é feito através de verificação formal usando o prove para garantir que todas as alternativas estão corretas e produzem o mesmo resultado.

Provas formais são úteis em áreas onde a precisão é crítica, como em sistemas de segurança e algoritmos de criptografia.
Utilizando bitvectors e provas formais, é possível garantir que as operações aritméticas e lógicas sejam realizadas corretamente sem falhas.

#Floating-Point Arithmetic (IEEE 754)
A aritmética de ponto flutuante no padrão IEEE 754 representa números reais com três componentes: sinal, expoente e mantissa. Usando essa representação, é possível lidar com uma vasta gama de valores, incluindo números muito grandes e muito pequenos. O padrão define formatos de precisão simples (32 bits) e dupla (64 bits). No entanto, operações de ponto flutuante podem introduzir erros de arredondamento e imprecisão devido à representação finita dos números.

##Exemplo 1:
Soma de Números de Ponto Flutuante

In [36]:
from cvc5.pythonic import *

# Definindo variáveis de ponto flutuante com precisão de 32 bits
x, y = FPs("x y", Float32())

# Definindo um solver para aritmética de ponto flutuante
s = SolverFor('QF_FP')

# Adicionando as restrições de soma e subtração
s.add(x + y == 10.5)
s.add(x - y == 2.5)

# Verificando satisfatibilidade
result = s.check()

if result == sat:
    m = s.model()
    print("result:", result)
    # Imprimindo os valores de x e y que satisfazem as restrições
    print("\n".join([str(k) + " : " + str(m[k]) for k in m]))
else:
    print("Não foi encontrada uma solução satisfatória.")




result: sat
x : 1.625*(2**2)
y : 1*(2**2)


##Exemplo 2:

In [37]:
from cvc5.pythonic import *

# Definindo variáveis de ponto flutuante com precisão de 32 bits
x, y = FPs("x y", Float32())

# Definindo um solver para aritmética de ponto flutuante
s = SolverFor('QF_FP')

# Adicionando as restrições de multiplicação e divisão
s.add(x * y == 12)
s.add(x / y == 3)

# Verificando satisfatibilidade
result = s.check()

if result == sat:
    m = s.model()
    print("result:", result)
    # Imprimindo os valores de x e y que satisfazem as restrições
    print("\n".join([str(k) + " : " + str(m[k]) for k in m]))
else:
    print("Não foi encontrada uma solução satisfatória.")


result: sat
x : -1.5*(2**2)
y : -1*(2**1)


A aritmética de ponto flutuante é um padrão essencial para cálculos numéricos em computadores. Ela permite representar números reais com precisão limitada, utilizando uma representação binária que simula números racionais. O padrão IEEE 754 define como números de ponto flutuante devem ser representados e operados em sistemas computacionais.

Principais utilidades:

* Alta precisão em cálculos científicos: Usado em áreas como física, astronomia e engenharia, onde a precisão dos cálculos numéricos é crucial.

* Processamento de sinais e imagens: A aritmética de ponto flutuante é útil para manipular grandes quantidades de dados de forma eficiente, como em computação gráfica, áudio e vídeo.

* Aprendizado de máquina: Muitos algoritmos de aprendizado de máquina e redes neurais fazem uso de cálculos em ponto flutuante para otimizar modelos e processar grandes volumes de dados.

* Simulações financeiras: No setor financeiro, a aritmética de ponto flutuante é usada para cálculos financeiros complexos que envolvem variáveis em grande escala.

#Strings/Sequents

##Exemplo 1: Verificar se uma string contém outra

In [38]:
from cvc5.pythonic import *

# Exemplo 1: Strings com operações de concatenação e comprimento
s1 = String("abc")
s2 = String("def")

# Concatenação
s3 = s1 + s2  # s3 será "abcdef"
print(f"Concatenação: {s3}")

# Comprimento de strings
len_s1 = Length(s1)  # len_s1 será 3
print(f"Comprimento de s1: {len_s1}")


# Exemplo 2: Usando Contains
s4 = String("Hello, world!")
s5 = String("world")

contains_result = Contains(s4, s5) # Verifica se s5 está contida em s4

# Create a solver and check for satisfiability
slv = Solver()
slv.add(contains_result)
result = slv.check()

# Print the result
print(f"'world' está em 'Hello, world!': {result == sat}")  # Output: True/False


# Exemplo 3:  Comparando strings
s6 = String("apple")
s7 = String("banana")

comparison_result = s6 == s7
print(f"'apple' é igual a 'banana': {simplify(comparison_result)}")


# Exemplo 4:  Substrings
s8 = String("This is a test string")
s9 = SubString(s8, 10, 3)  # Extrai "tes" a partir da posição 10 com comprimento 3
print(f"Substring de s8: {simplify(s9)}")


# Exemplo 5:  Usando um Solver para Strings
s = SolverFor('QF_SLIA') # Solver para strings

str_var1 = String("x")
str_var2 = String("y")
str_var3 = String("z")

# Restrições
s.add(Length(str_var1) == 1)
s.add(str_var3 == str_var1 + str_var2)
s.add(Length(str_var3) == 2)
s.add(Contains(str_var3, String("a")))  # 'a' está em str_var3

# Check for satisfiability
result = s.check()
print(f"Satisfatibilidade: {result}")

if result == sat:
  m = s.model()
  print(f"Modelo: str_var1 = {m[str_var1]}, str_var2 = {m[str_var2]}, str_var3 = {m[str_var3]}")

Concatenação: +(abc, def)
Comprimento de s1: Length(abc)
'world' está em 'Hello, world!': True
'apple' é igual a 'banana': apple == banana
Substring de s8: SubString(|This is a test string|, 10, 3)
Satisfatibilidade: sat
Modelo: str_var1 = "B", str_var2 = "A", str_var3 = "BA"


In [39]:
from cvc5.pythonic import *
p, x, i = {}, {}, {}
for k in range(1, 13): p[k] = Bool("p" + str(k))
for k in range(1, 6): x[k] = String("x" + str(k))
for k in range(1,7): i[k] = String("i" + str(k))

result = StringVal("abbaabb")

s = SolverFor('QF_SLIA')
s.add(And(Length(x[1]) <= 2, Length(x[2]) <= 2))

s.add(i[1] == If(p[1], x[1], x[2]))
s.add(i[2] == If(p[2], x[1], x[2]))
s.add(x[3] == Concat(i[1], i[2]))

s.add(i[3] == If(p[3], x[1], If(p[4], x[2], x[3])))
s.add(i[4] == If(p[5], x[1], If(p[6], x[2], x[3])))
s.add(x[4] == Concat(i[3], i[4]))

s.add(i[5] == If(p[7], x[1], If(p[8], x[2], If(p[9], x[3], x[4]))))
s.add(i[6] == If(p[10], x[1], If(p[11], x[2], If(p[12], x[3], x[4]))))
s.add(x[5] == Concat(i[5], i[6]))

s.add(x[5] == result)

print(s.model() if s.check() == sat else "unsat")

[i1 = "a", i2 = "bb", i3 = "a", i4 = "abb", i5 = "abb", i6 = "aabb", p1 = True, p10 = False, p11 = False, p12 = False, p2 = False, p3 = True, p4 = True, p5 = False, p6 = False, p7 = False, p8 = False, p9 = True, x1 = "a", x2 = "bb", x3 = "abb", x4 = "aabb", x5 = "abbaabb"]


In [40]:
from cvc5.pythonic import *
x, y, z = Consts("x y z", SeqSort(IntSort()))

s = SolverFor('QF_SLIA')

s.add(Length(x) > 0)
s.add(x[0] + x[Length(x) - 1] == 9)
s.add(y == Concat(x,x))
s.add(z == Concat(Unit(IntVal(3)), Unit(IntVal(4)), Unit(IntVal(5))))
s.add(Contains(y, z))

print(s.model() if s.check() == sat else "unsat")


[x = (seq.++ (seq.unit 4) (seq.unit 3) (seq.unit 4) (seq.unit 5))(), y = (seq.++ (seq.unit 4) (seq.unit 3) (seq.unit 4) (seq.unit 5) (seq.unit 4) (seq.unit 3) (seq.unit 4) (seq.unit 5))(), z = (seq.++ (seq.unit 3) (seq.unit 4) (seq.unit 5))()]


A manipulação de strings e sequências com solvers SMT oferece um poder significativo em termos de verificação automática e análise simbólica. Isso é útil em uma ampla gama de áreas, incluindo verificação de programas, segurança de software, processamento de texto e modelagem de estruturas de dados complexas. A capacidade de usar um solver para verificar propriedades sobre strings e sequências sem precisar gerar instâncias concretas é um poderoso recurso, especialmente em contextos de verificação formal e análise lógica.

# Algebraic Datatypes

In [41]:
from cvc5.pythonic import *
decl = Datatype("tree")
decl.declare("node", ("data", IntSort()), ("left", decl), ("right", decl))
decl.declare("nil")
Tree = decl.create()

x, y = Consts("x y", Tree)

s = SolverFor('ALL')
s.add(Tree.is_node(x))
s.add(Tree.is_node(y))
s.add(Tree.left(x) == Tree.right(y))
s.add(Tree.data(x) > 100)

print(s.model() if s.check() == sat else "unsat")




[x = node(101, nil, node(0, nil, nil)), y = node(0, node(0, nil, node(0, nil, nil)), nil)]


Tipos Algébricos Recursivos: Ele demonstra como trabalhar com tipos recursivos, que são fundamentais em muitas linguagens funcionais e em problemas matemáticos complexos, como a manipulação de estruturas de dados dinâmicas, como árvores e listas encadeadas.

Verificação Formal: Ao usar o solver cvc5, o código permite a verificação formal de propriedades sobre essas estruturas de dados. Em vez de apenas criar e manipular árvores de forma imperativa, o código formaliza propriedades como simetria (subárvores) e valores (dados em um nó) e usa um solver para garantir que essas propriedades sejam atendidas.

Aplicações em Inteligência Artificial e Lógica: Esse tipo de modelagem é útil em contextos de verificação automática, inteligência artificial (IA) e soluções de problemas de lógica. Por exemplo, pode ser utilizado em:

Verificação de programas: Garantir que um programa manipule árvores de forma correta e atenda a condições especificadas.
Lógica simbólica: Resolver problemas lógicos envolvendo estruturas de dados complexas.
Árvores de decisão e algoritmos recursivos: Modelar e validar propriedades sobre algoritmos que operam em árvores.


#Sets/Relations

In [42]:
from cvc5.pythonic import *
S = DeclareSort("S")
A, B, C = [Set(i, S) for i in ["A","B","C"]]
s = SolverFor('QF_FS')
s.add(Not((A | (B & C)) == ((A | B) & (A | C))))
print(s.check())


unsat


O uso de sets e relações em SMT tem várias aplicações práticas:

Verificação de Programas: Verificar se operações de conjuntos (como união e interseção) estão sendo feitas corretamente dentro de um programa, garantindo que o comportamento esperado seja mantido.

Sistemas de Banco de Dados: Modelar operações sobre conjuntos de dados, como a união de tabelas ou a interseção de conjuntos de registros.

Algoritmos de Computação Combinatória: Resolver problemas de algoritmos combinatórios que envolvem operações sobre conjuntos, como problemas de agrupamento, particionamento de conjuntos e teoria dos grafos.

Teoria da Computação e Lógicas Formais: A análise algébrica e lógica de relações e conjuntos é uma parte fundamental da teoria da computação, especialmente em contextos de lógicas formais e prova de teoremas.

Segurança e Criptografia: Em protocolos de segurança, relações e operações de conjuntos podem ser usadas para modelar permissões de acesso, criptografia de dados ou qualquer operação envolvendo agregação ou seleção de dados.

Conclusão
A manipulação de sets e relações no contexto de solvers SMT oferece uma maneira formal e eficiente de verificar propriedades algébricas e lógicas sobre dados, o que é essencial para diversas áreas da computação, incluindo verificação de programas, teoria da computação, segurança, bancos de dados, e algoritmos de otimização. O uso de SMT para verificar a satisfatibilidade de expressões envolvendo conjuntos e relações pode ser uma ferramenta poderosa em modelagem formal e análise simbólica.

#Quantifiers

In [43]:
from cvc5.pythonic import *
s = SolverFor('UF')
s.setOption('enum-inst', True)
s.setOption('finite-model-find', True)
s.setOption('mbqi', True)
s.setOption('sygus-inst', True)

In [44]:
from cvc5.pythonic import *
S = DeclareSort("S")
Bool = BoolSort()
Human = Function("Human", S, Bool)
Mortal = Function("Mortal", S, Bool)
Socrates = Const("Socrates", S)

s = SolverFor('UF')

x = Const("x", S)
s.add(ForAll([x], Implies(Human(x), Mortal(x))))
s.add(Human(Socrates))
s.add(Not(Mortal(Socrates)))

print(s.check())


unsat


A importância dos quantificadores em lógica de primeira ordem:
Quantificadores Universais (ForAll): O quantificador ForAll é utilizado para expressar propriedades gerais que se aplicam a todos os elementos de um determinado domínio. No seu exemplo, ele é usado para afirmar que todos os seres humanos são mortais.

Quantificadores Existenciais (Exists): Embora não tenha sido utilizado explicitamente no seu código, o quantificador existencial (Exists) é usado para afirmar que existe pelo menos um elemento em um domínio que satisfaz uma certa condição. Por exemplo, você poderia afirmar que "existe um ser humano que é imortal", utilizando Exists([x], And(Human(x), Not(Mortal(x)))).

Verificação de Propriedades Lógicas: O uso de quantificadores permite que você escreva e verifique propriedades lógicas mais complexas que envolvem todo um domínio ou a existência de elementos específicos que satisfazem certas condições. Eles são fundamentais em provas formais e verificação automática de teoremas, sendo usados para modelar e verificar especificações de sistemas ou afirmações lógicas.

Aplicações em Inteligência Artificial e Computação: Os quantificadores são utilizados em vários contextos, como:

Verificação Formal: Garantir que um sistema de software ou hardware satisfaça as propriedades desejadas em todos os casos possíveis.
Lógica de Programação: Especificação de invariantes de loops e verificação de correção de programas.
Raciocínio Automático e Inteligência Artificial: Em sistemas baseados em conhecimento (como ontologias e sistemas de raciocínio), os quantificadores são usados para inferir novos conhecimentos a partir de dados existentes.
Conclusão
A utilização de quantificadores universais e existenciais permite expressar propriedades sobre todos os elementos de um domínio ou a existência de certos elementos que satisfaçam determinadas condições. No seu exemplo, o uso de ForAll para afirmar que todos os humanos são mortais e a contradição gerada por afirmar que Socrates não é mortal ajuda a entender a importância dos quantificadores para expressar e verificar propriedades lógicas complexas em sistemas formais.

#Separation Logic

In [45]:
from cvc5.pythonic import *

# Create solver with separation logic
solver = SolverFor("ALL")

# Declare a custom heap type (Int -> Int)
Heap = Datatype("Heap")
Heap.declare("pto", ("addr", IntSort()), ("value", IntSort()))  # Represents address -> value mapping
Heap.create()

# Declare constants for heap-like behavior
x, a, b = Ints('x a b')

# Create a custom function to represent heap allocations
pto = Function("pto", IntSort(), IntSort(), BoolSort())  # A function to represent heap allocations (pointer-to)

# Assert conflicting heap allocations
solver.add(And(pto(x, a), pto(x, b)))  # x points to a and x points to b
solver.add(a != b)  # a and b are not equal

# Check for satisfiability
print(solver.check())  # Output: unsat



sat


In [46]:
from cvc5.pythonic import *

# Create solver
solver = SolverFor("ALL")

# Declare custom sort U
U = DeclareSort('U')

# Declare custom heap type (U -> Int)
Heap = Datatype("Heap")
Heap.declare("pto", ("addr", U), ("value", IntSort()))  # Represents addr -> value
Heap.create()

# Declare constants
x = Const('x', U)
a = Int('a')

# Create a custom function to represent heap allocations (Pto)
pto = Function("pto", U, IntSort(), BoolSort())

# Assert non-empty heap (x points to a value a)
solver.add(pto(x, a))  # Represent that x points to a

# Check for satisfiability
print(solver.check())  # Expected output: sat

sat


In [47]:
from cvc5.pythonic import *

# Create solver
solver = SolverFor("ALL")

# Declare Node datatype
Node = Datatype('Node')
Node.declare('node', ('data', IntSort()),
                   ('left', IntSort()),
                   ('right', IntSort()))
Node = Node.create()  # This already represents a sort

# Declare a function to represent heap: Int -> Node
Heap = Function('Heap', IntSort(), Node)  # No need for .sort()

# Declare constants
x, y, z = Ints('x y z')

# Create a node instance
node_x = Node.node(0, y, z)

# Assert heap allocation (Heap(x) == node_x)
solver.add(Heap(x) == node_x)

# Check for satisfiability
print(solver.check())  # Expected output: sat


sat


A Separation Logic é uma ferramenta poderosa para raciocinar sobre alocação dinâmica, prevenir bugs de memória e analisar programas concorrentes de forma eficiente. Seu uso em SMT solvers como cvc5 permite automatizar a verificação de código e garantir maior segurança e correção em software crítico.

Principais Benefícios da Separation Logic

Raciocínio Local
Permite raciocinar sobre partes específicas da memória sem precisar considerar o estado global inteiro.
Exemplo: Se sabemos que um ponteiro p aponta para um nó de uma lista, podemos provar propriedades sobre ele sem precisar analisar toda a lista.

Evita Condições de Corrupção de Memória
Modela explicitamente a alocação e liberação de memória, ajudando a detectar:
Derefências inválidas (accesso a ponteiros inválidos)
Vazamentos de memória (memória não liberada)
Uso de memória desalocada (use-after-free)

Prova Propriedades de Estruturas de Dados Dinâmicas
Útil para provar correção de estruturas como:
Listas encadeadas, árvores, heaps.
Exemplo: Pode garantir que um nó de uma árvore não tem filhos duplicados ou que uma lista não tem ciclos.

Facilita Análise de Programas Concorrentes
A lógica de separação permite definir partições disjuntas de memória, o que é essencial para programação concorrente segura.
Ajuda a evitar problemas como condições de corrida e deadlocks.
🔹 Como Funciona na Prática?
A lógica de separação introduz dois principais operadores:

Separating Conjunction ( ∗ )

Expressa que duas partes da memória são separadas e independentes.
Exemplo:
𝑃
(
𝑥
)
∗
𝑄
(
𝑦
)
P(x)∗Q(y)
significa que P(x) e Q(y) ocupam regiões disjuntas da memória.
Separating Implication ( −∗ )

Expressa que se tivermos P, então podemos transformar P em Q, preservando separação.
Útil para especificar condições de pré/pós-condição em funções.